File: api/models/customer.py
"""
Customer model – represents individual or business clients.

Schema reference: docs/ARCHITECT.md § Model: Customer
Dictionary:       docs/DICTIONARY.md  (balance = công nợ)

Balance semantics
-----------------
  balance > 0  →  Customer owes money  (debt / công nợ dương)
  balance < 0  →  Customer has credit  (over-paid / công nợ âm)
  balance == 0 →  Settled
"""

import uuid
from typing import TYPE_CHECKING, List, Optional

from sqlmodel import Field, Relationship, SQLModel

from .enums import CustomerType

if TYPE_CHECKING:
    from .invoice import Invoice
    from .quote import Quote
    from .ticket import Ticket
    from .transaction import Transaction

# ---------------------------------------------------------------------------
# Shared / base properties
# ---------------------------------------------------------------------------

class CustomerBase(SQLModel):
    """Fields shared between create/read schemas and the DB table."""

    name: str = Field(
        index=True,
        min_length=1,
        max_length=255,
        description="Full name of the individual or registered business name.",
    )
    email: Optional[str] = Field(
        default=None,
        index=True,
        unique=True,
        max_length=255,
        description="Optional customer email used for contact and invoice delivery.",
    )
    phone: Optional[str] = Field(
        default=None,
        max_length=30,
        description="Optional customer phone number.",
    )
    address: Optional[str] = Field(
        default=None,
        max_length=500,
        description="Optional billing/customer address.",
    )
    tax_code: Optional[str] = Field(
        default=None,
        index=True,
        unique=True,
        max_length=100,
        description="Optional Vietnamese tax code for invoicing.",
    )
    type: CustomerType = Field(
        default=CustomerType.INDIVIDUAL,
        description="INDIVIDUAL for personal travellers; BUSINESS for corporate accounts.",
    )
    balance: float = Field(
        default=0.0,
        description=(
            "Current debt balance (công nợ). "
            "Positive = customer owes; Negative = customer has credit."
        ),
    )
    is_active: bool = Field(
        default=True,
        description="Soft-archive flag. Inactive customers stay in history but should be treated as archived in the UI.",
    )


# ---------------------------------------------------------------------------
# DB Table
# ---------------------------------------------------------------------------

class Customer(CustomerBase, table=True):
    """Persisted customer entity stored in the `customer` table."""

    __tablename__ = "customer"

    id: Optional[uuid.UUID] = Field(
        default_factory=uuid.uuid4,
        primary_key=True,
        index=True,
        nullable=False,
    )

    # Relationships – populated lazily by SQLAlchemy; not included in API responses by default.
    tickets: List["Ticket"] = Relationship(back_populates="customer")
    transactions: List["Transaction"] = Relationship(back_populates="customer")
    invoices: List["Invoice"] = Relationship(back_populates="customer")
    quotes: List["Quote"] = Relationship(back_populates="customer")


# ---------------------------------------------------------------------------
# Request / Response schemas
# ---------------------------------------------------------------------------

class CustomerCreate(CustomerBase):
    """Payload accepted by POST /customers."""
    pass


class CustomerRead(CustomerBase):
    """Public representation of a customer."""

    id: uuid.UUID


class CustomerDirectoryItem(SQLModel):
    """Slim customer payload used by the directory page."""

    id: uuid.UUID
    full_name: str
    phone: Optional[str] = None
    current_balance: float
    is_active: bool = True


class CustomerUpdate(SQLModel):
    """All fields optional for partial PATCH payloads."""

    name: Optional[str] = Field(default=None, min_length=1, max_length=255)
    email: Optional[str] = Field(default=None, max_length=255)
    phone: Optional[str] = Field(default=None, max_length=30)
    address: Optional[str] = Field(default=None, max_length=500)
    tax_code: Optional[str] = Field(default=None, max_length=100)
    type: Optional[CustomerType] = None
    balance: Optional[float] = None
    is_active: Optional[bool] = None

File: web/src/schemas/customer.ts
/**
 * customer.ts – Zod schemas for the Customer domain.
 *
 * Mirrors api/models/customer.py (CustomerCreate, CustomerRead, CustomerUpdate).
 *
 * Balance semantics (công nợ) – from docs/DICTIONARY.md:
 *   balance > 0  →  Customer owes money  (debt)
 *   balance < 0  →  Customer has credit  (over-paid)
 *   balance === 0 → Settled
 */

import { z } from "zod";
import { CustomerTypeSchema } from "./enums";
import { TicketReadSchema } from "./ticket";
import { TransactionReadSchema } from "./transaction";

// ---------------------------------------------------------------------------
// Shared base
// ---------------------------------------------------------------------------

const CustomerBaseSchema = z.object({
  /** Full name of the individual or registered business. Maps to Python: name */
  name: z
    .string()
    .min(1, "Name is required.")
    .max(255, "Name must be at most 255 characters."),

  /** Account category. Maps to Python: type */
  type: CustomerTypeSchema.default("INDIVIDUAL"),

  /**
   * Current debt balance (công nợ).
   * Positive = customer owes; Negative = customer has credit.
   * Maps to Python: balance
   */
  balance: z.number().default(0),

  /** Soft-archive flag. Archived customers remain in history. */
  is_active: z.boolean().default(true),

  /** Optional customer email used for billing and contact. */
  email: z.string().email().max(255).nullable().optional(),

  /** Optional customer phone number. */
  phone: z.string().max(30).nullable().optional(),

  /** Optional billing address. */
  address: z.string().max(500).nullable().optional(),

  /** Optional tax code for invoice issuance. */
  tax_code: z.string().max(100).nullable().optional(),
});

// ---------------------------------------------------------------------------
// CustomerCreate – POST /customers
// ---------------------------------------------------------------------------

export const CustomerCreateSchema = CustomerBaseSchema;
export type CustomerCreate = z.infer<typeof CustomerCreateSchema>;

// ---------------------------------------------------------------------------
// CustomerRead – GET /customers/:id
// ---------------------------------------------------------------------------

export const CustomerReadSchema = CustomerBaseSchema.extend({
  /** UUID assigned by the backend. Read-only on the frontend. */
  id: z.string().uuid(),
});

export type CustomerRead = z.infer<typeof CustomerReadSchema>;

// ---------------------------------------------------------------------------
// CustomerUpdate – PATCH /customers/:id (all fields optional)
// ---------------------------------------------------------------------------

export const CustomerUpdateSchema = z.object({
  name: z
    .string()
    .min(1, "Name is required.")
    .max(255, "Name must be at most 255 characters.")
    .optional(),

  type: CustomerTypeSchema.optional(),

  /** Direct balance override – normally mutated via Transactions only. */
  balance: z.number().optional(),

  is_active: z.boolean().optional(),
  email: z.string().email().max(255).nullable().optional(),
  phone: z.string().max(30).nullable().optional(),
  address: z.string().max(500).nullable().optional(),
  tax_code: z.string().max(100).nullable().optional(),
});

export type CustomerUpdate = z.infer<typeof CustomerUpdateSchema>;

export const BalanceStateSchema = z.enum(["debt", "settled", "credit"]);
export type BalanceState = z.infer<typeof BalanceStateSchema>;

export const LedgerEntrySchema = z.object({
  id: z.string().uuid(),
  entry_type: z.enum(["ticket", "payment", "adjustment"]),
  created_at: z.coerce.date(),
  content: z.string(),
  amount: z.number(),
  running_balance: z.number(),
  ticket: TicketReadSchema.nullable().optional(),
  transaction: TransactionReadSchema.nullable().optional(),
});

export type LedgerEntry = z.infer<typeof LedgerEntrySchema>;

export const CustomerLedgerSchema = z.object({
  customer: CustomerReadSchema,
  current_balance: z.number(),
  balance_state: BalanceStateSchema,
  entries: z.array(LedgerEntrySchema),
});

export type CustomerLedger = z.infer<typeof CustomerLedgerSchema>;

export const RecordPaymentSchema = z.object({
  amount: z.number().positive("Amount must be a positive number."),
  method: z.string().min(1, "Method is required."),
  note: z.string().min(1, "Note is required.").max(2000, "Note must be at most 2000 characters."),
  evidence_url: z
    .string()
    .url("evidence_url must be a valid URL.")
    .max(2048)
    .nullable()
    .optional(),
  linked_ticket_id: z.string().uuid().optional(),
});

export type RecordPayment = z.infer<typeof RecordPaymentSchema>;

export const CustomerDirectoryItemSchema = z.object({
  id: z.string().uuid(),
  full_name: z.string(),
  phone: z.string().nullable().optional(),
  current_balance: z.number(),
  is_active: z.boolean().default(true),
});

export type CustomerDirectoryItem = z.infer<typeof CustomerDirectoryItemSchema>;
